### PR TITLE
Add "efficient_finetuning" flag and supports bit_fit and norm_fit

### PR DESCRIPTION
**Pull request by sxjscience**
*Description of changes:*
- add "optimization.efficient_finetune" flag. We are considering to support a bunch of efficient finetuning algorithms. For now, we support
   - "bit_fit": "BitFit: Simple Parameter-efficient Fine-tuning for Transformer-based Masked Language-models" (https://arxiv.org/abs/2106.10199)
   - "norm_fit": This is an extension of the algorithm in "Training BatchNorm and Only BatchNorm: On the Expressive Power of Random Features in CNNs" (https://arxiv.org/abs/2003.00152) and BitFit. We finetune both the parameters in the norm layers as long as the biases. Actually, it may make more sense for modern NN architectures because we are removing the bias terms.
- Fix the bug that the categorical label won't be transformed to the original space.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 1]
```diff
@@ -211,11 +211,34 @@ def get_weight_decay_param_names(model: nn.Module):
     -------
     A list of parameter names not using weight decay.
     """
-    decay_param_names = get_parameter_names(model, [nn.LayerNorm])
+    # By default, we should not apply weight decay for all the norm layers
+    decay_param_names = get_parameter_names(model,
+                                            [nn.LayerNorm, nn.BatchNorm1d, nn.BatchNorm2d, nn.BatchNorm3d,
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/utils.py:216]: Besides layer norm, do we have references about not using weight decay on other norms?

**Comment by szha** [comment]
Job PR-1666-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1666/6/index.html

File: text/src/autogluon/text/automm/optimization/utils.py (hunk 1 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/optimization/utils.py:216]: In https://arxiv.org/pdf/1812.01187.pdf (Section "No bias decay"), the author found that do not add wd to \gamma and \beta in BN layers will improve the performance.

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 2]
```diff
@@ -211,11 +211,34 @@ def get_weight_decay_param_names(model: nn.Module):
...
+    ----------
+    model
+        A Pytorch model
+
+    Returns
+    -------
+    norm_param_names
+        A list of names that
+    """
+    all_param_names = get_parameter_names(model, [])
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/utils.py:235]: `all_param_names = [name for name, _ in model.named_parameters()]` seems more straightforward.

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 3]
```diff
@@ -346,6 +369,7 @@ def apply_layerwise_lr_decay(
         lr: float,
         lr_decay: float,
         weight_decay: float,
+        efficient_finetune: Optional[str] = None,
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/utils.py:372]: Do we also need to add this option for `apply_single_lr` and `apply_two_stages_lr`? Do the papers use layer-wise learning rate decay in efficient fine-tuning?

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 4]
```diff
@@ -373,8 +399,17 @@ def apply_layerwise_lr_decay(
...
     for name, param in model.named_parameters():
+        if efficient_finetune == 'bit_fit':
+            # For bit_fit, we disable tuning everything except the bias terms
+            if 'bias' not in name:
+                param.requires_grad = False
+        elif efficient_finetune == 'norm_fit':
+            # For norm-fit, we finetune all the normalization layers and bias layers
+            if name not in norm_param_names and 'bias' not in name:
+                param.requires_grad = False
+
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/utils.py:412]: If the user provides a string not in [bit_fit, norm_fit], we need to raise errors, like how we did in the top-k average choices.

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 5]
```diff
@@ -211,11 +211,34 @@ def get_weight_decay_param_names(model: nn.Module):
...
+    # By default, we should not apply weight decay for all the norm layers
+    decay_param_names = get_parameter_names(model,
+                                            [nn.LayerNorm, nn.BatchNorm1d, nn.BatchNorm2d, nn.BatchNorm3d,
+                                             nn.GroupNorm])
     decay_param_names = [name for name in decay_param_names if "bias" not in name]
     return decay_param_names
 
 
+def get_norm_layer_param_names(model: nn.Module):
+    """Get parameters associated with the normalization layers
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/utils.py:223]: To make format consistent, add a newline after `"""`.

**Comment by sxjscience** [comment]
I've conducted the initial benchmark and noticed that "norm_fit" outperforms "bit_fit". Thus, we can recommend users to use "norm_fit" for efficient finetuning. The number of trainable parameters for norm_fit will be very close to that of "bit_fit" but the performance can be better.

File: text/src/autogluon/text/automm/optimization/utils.py (hunk 4 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/optimization/utils.py:412]: Should be fine for now. We may still give user a fallback solution.

File: text/src/autogluon/text/automm/optimization/utils.py (hunk 3 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/optimization/utils.py:372]: No. The original paper hasn't combined bit_fit with layerwise lr decay. Also, should we remove apply_single_lr in the future?

**Comment by szha** [comment]
Job PR-1666-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1666/8/index.html

**Comment by sxjscience** [comment]
To provide more details, on three multilingual datasets, "norm_fit" is 0.3% slower than "bit_fit" in terms of training speed but can consistently outperform bit_fit in terms of the final performance of the model, in both the monolingual and cross-lingual transfer setting.

File: text/src/autogluon/text/automm/optimization/utils.py (hunk 3 above)
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/optimization/utils.py:373]: Since the original paper didn't use layer-wise lr decay, we may need to test whether layer-wise lr decay is a better choice than single lr for bit_fit.

**Comment by szha** [comment]
Job PR-1666-11 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1666/11/index.html

File: text/src/autogluon/text/automm/optimization/utils.py (hunk 5 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/optimization/utils.py]: revised.

File: text/src/autogluon/text/automm/optimization/utils.py (hunk 3 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/optimization/utils.py:373]: After additional experiments, I noticed that lr_decay=1.0 gives the best result. Thus, we may stick to lr_decay=1.0 when applying efficient_finetune. In addition, we may only implement that for `apply_layerwise_lr_decay` because `lr_decay=1.0` is equivalent to `apply_single_lr`.

**Comment by sxjscience** [comment]
Resolved the comments @zhiqiangdon

File: text/src/autogluon/text/automm/data/preprocess_dataframe.py [hunk 6]
```diff
@@ -435,6 +436,8 @@ def transform_prediction(
         ----------
         y_pred
             The model's output logits.
+        inverse_categorical
+            Whether to transformer categorical value back to the original space, e.g., string.
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/data/preprocess_dataframe.py:440]: transformer -> transform

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 7]
```diff
@@ -211,11 +211,35 @@ def get_weight_decay_param_names(model: nn.Module):
...
+
+    Parameters
+    ----------
+    model
+        A Pytorch model
+
+    Returns
+    -------
+    norm_param_names
+        A list of names of the parameters that are either part of normalization layer or bias terms.
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/utils.py:234]: Should be `A list of normalization parameter names`.

File: text/src/autogluon/text/automm/optimization/utils.py (hunk 3 above)
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/optimization/utils.py:373]: We can remove `apply_single_lr` later. If a user wants to use efficient fine-tuning, the user needs to customize both `efficient_finetune ` and `lr_decay `. We may need a tutorial for configuring efficient fine-tuning.

File: text/src/autogluon/text/automm/optimization/utils.py (hunk 7 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/optimization/utils.py]: revised.

File: text/src/autogluon/text/automm/data/preprocess_dataframe.py (hunk 6 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/data/preprocess_dataframe.py]: revised.

**Comment by szha** [comment]
Job PR-1666-15 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1666/15/index.html

File: text/src/autogluon/text/automm/configs/optimization/adamw.yaml [hunk 8]
```diff
@@ -15,3 +15,4 @@ optimization:
   top_k: 3
   top_k_average_method: "greedy_soup"  # We support averaging method described in https://arxiv.org/pdf/2203.05482.pdf.
                                        # Currently support "uniform_soup", "greedy_soup", and "best".
+  efficient_finetune: null  # Can be 'bit_fit' (only finetune bias), 'norm_fit' (finetune the normalization terms + bias terms), or null
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/configs/optimization/adamw.yaml:18]: We may need to add the comment: set `lr_decay`= 1.0 if using `bit_fit` or `norm_fit`.

**Review by zhiqiangdon — APPROVED** [review]

File: text/src/autogluon/text/automm/configs/optimization/adamw.yaml (hunk 8 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/configs/optimization/adamw.yaml:18]: Yes, I will later work on a notebook about efficient finetuning. The plan is to also add adapter (potentially other methods) and compare a few options.
